csharp/test/TestCommon/TestHelper.cs

```diff
@@ -284,3 +284,260 @@ public static async Task<int> RunTestAsync<T>(string[] args) where T : TestHelpe
+                return sb.ToString();
+            }
+        }
+
+        public static string GetTestHost(Dictionary<string, string> properties)
+        {
+            if (!properties.TryGetValue("Test.Host", out string? host))
+            {
+                host = "127.0.0.1";
+            }
+            return host;
+        }
+
+        public static Protocol GetTestProtocol(Dictionary<string, string> properties)
+        {
+            if (!properties.TryGetValue("Test.Protocol", out string? value))
+            {
+                return Protocol.Ice2;
+            }
+            try
+            {
+                return ProtocolExtensions.Parse(value);
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidConfigurationException(
+                    $"invalid value for for Test.Protocol: `{value}'", ex);
+            }
+        }
+
+        public static ZeroC.Ice.Encoding GetTestEncoding(Dictionary<string, string> properties)
+            => ProtocolExtensions.GetEncoding(GetTestProtocol(properties));
+
+        public static string GetTestTransport(Dictionary<string, string> properties)
+        {
+            if (properties.TryGetValue("Test.Transport", out string? transport))
+            {
+                if (GetTestProtocol(properties) == Protocol.Ice1)
+                {
+                    return transport;
+                }
+                return transport switch
+                {
+                    "wss" => "ws",
+                    "ssl" => "tcp",
+                    _ => transport,
+                };
+            }
+            else
+            {
+                return "tcp";
+            }
+        }
+
+        public static ushort GetTestBasePort(Dictionary<string, string> properties)
+        {
+            ushort basePort = 12010;
+            if (properties.TryGetValue("Test.BasePort", out string? value))
+            {
+                basePort = ushort.Parse(value);
+            }
+            return basePort;
+        }
+
+        public static Dictionary<string, string> CreateTestProperties(
+            ref string[] args,
+            Dictionary<string, string>? defaults = null)
+        {
+            var properties =
+                defaults == null ? new Dictionary<string, string>() : new Dictionary<string, string>(defaults);
+            properties.ParseIceArgs(ref args);
+            properties.ParseArgs(ref args, "Test");
+            return properties;
+        }
+
+        public string GetTestProxy(string identity, int num = 0, string? transport = null) =>
+            GetTestProxy(identity, Communicator.GetProperties(), num, transport);
+
+        public abstract Task RunAsync(string[] args);
+
+        public static async Task<int> RunTestAsync<T>(Communicator communicator, string[] args)
+            where T : TestHelper, new()
+        {
+            try
+            {
+                var helper = new T() { Communicator = communicator };
```

Comment:
That's nice, I was surprised when I see the init method was private, and the abstract class has no constructor.

Reply:
What is less nice is:
```
public Communicator Communicator { get; private init; } = null!;
```
without this ` = null!` we get a warning.

Reply:
is it the same if you make it `protected init`?

Reply:
Yes.